### ISSUE TITLE
fix font for Kyrgyz

### ISSUE THREAD

**Issue by jonorthwash** [common-voice/common-voice#1141]
The letters ө, ү, and ң are different heights from other letters.  They should have the same dimensions as о, у, and н, respectively.  A different default font probably needs to be selected, or a web font should be used.

See example below:

![2018-06-23-14 03 32_001](https://user-images.githubusercontent.com/963849/41812293-5788ff26-76ee-11e8-92df-533891f35aba.png)

**Comment by ftyers** [comment]
This issue also effects the ҫ in Chuvash:

![captura de 2018-06-23 20-07-58](https://user-images.githubusercontent.com/449545/41812320-29419b2c-7721-11e8-9f1d-c17df1898563.png)

**Comment by jonorthwash** [comment]
For me, the Chuvash ӑ, ӗ, and ҫ all show up *larger* than other letters:
![2018-06-24-23 17 34_001](https://user-images.githubusercontent.com/963849/41828717-d219c768-7804-11e8-8ac4-8a5c24ec0871.png)

This is in chromium on linux.

**Comment by Gregoor** [comment]
In the new contribution expeirence we're using the OpenSans font. It doesn't seem to have that problem, does it?
https://voice.allizom.org/cv/speak

**Comment by jonorthwash** [comment]
The problem isn't there in the sentences presented in the new contribution experience, but there are still some font issues:

In the following, the "ү" and "ө" characters in "Сүйлө" are the wrong height:
![2018-06-30-02 47 23_001](https://user-images.githubusercontent.com/963849/42122449-06ea51ce-7c10-11e8-85e8-a7695ce9ae54.png)

It'll probably be a problem for Chuvash too once "listen" is localised.

**Comment by jonorthwash** [comment]
The new interface also has an issue here (note height of ү and ң):

![2018-07-01-14 49 53_001](https://user-images.githubusercontent.com/963849/42137652-36b07918-7d3e-11e8-8e34-bc70d678727b.png)

**Comment by Gregoor** [comment]
So I think this might be an issue with the font we're using ZillaSlab. This is a cyrillic alphabet, no? Maybe we should ask it over here as well: https://github.com/mozilla/zilla-slab/issues/17

Otherwise we'd need to use a different font, wdyt @m-branson?

**Comment by mikehenrty** [comment]
> Maybe we should ask it over here as well: mozilla/zilla-slab#17

Great idea! I did it

**Comment by mbransn** [comment]
If languages are having display issues with Zilla Slab, the immediate recommendation is to attempt styling in Open Sans. If that doesn't rectify the issue, we'll need to confirm different web fonts for those languages.

Has anyone confirmed that this is still an issue for the languages presented here, Kyrgyz and Chuvash? @jonorthwash @ftyers

**Comment by jonorthwash** [comment]
Oof, still a problem.
![Screenshot from 2020-03-09 21-13-57 cropped](https://user-images.githubusercontent.com/963849/76270301-28838a80-624b-11ea-9906-715bfe40a1ae.png)

@mbransn, how can I attempt styling in Open Sans?  I assume something like Tampermonkey is the way to go?

**Comment by phirework** [comment]
@jonorthwash If you use your browser's inspector tool to open up the CSS, you can see how it looks with different font families. Here's a guide for how to do that: https://developer.mozilla.org/en-US/docs/Tools/Page_Inspector/How_to/Edit_fonts

**Comment by jonorthwash** [comment]
@phirework, yeah, I know how to play with pages' CSS in my browser's inspector tool.  But I'm not going to do this every time I visit the page, and I doubt most users will have the patience (or even comfort with doing this) to do this.

I guess I'm not totally sure what's being asked of me here....

**Comment by mbransn** [comment]
> @mbransn, how can I attempt styling in Open Sans? I assume something like Tampermonkey is the way to go?

@jonorthwash looks to me like @phirework was making a recommendation on how to accomplish your question. Not a recommendation for every visity, merely a request to help test if Open Sans works for Kyrgyz character presentation.

**Comment by jonorthwash** [comment]
> a request to help test if Open Sans works for Kyrgyz character presentation.

Ah, I thought it was known that Open Sans has good support for Cyrillic (it does: it supports all of the Cyrillic unicode block, although none of the Cyrillic Supplement unicode block).

I'm still confused about why I'm being asked to offer confirmation about general properties of a widely available and well documented font that I've never contributed to.  Or am I misunderstanding some pragmatics here?

**Comment by jonorthwash** [comment]
I should also point out, in case it was missed, that most of the text in my screenshot was in Open Sans and looks fine.

**Comment by jonorthwash** [comment]
Also, the lack of good Cyrillic support in Zilla Slab appears to be a [known issue](https://github.com/mozilla/zilla-slab/issues/37).

**Comment by jonorthwash** [comment]
On the other hand, Chuvash looks fairly good in Zilla Slab now.  The places where Open Sans is used for Chuvash can probably be switched back to Zilla Slab.

**Comment by phirework** [comment]
Apologies for the miscommunication @jonorthwash, I did indeed misunderstand your question about styling it with Open Sans. I've been looking into this issue and I haven't been able to replicate what it looks like in your screenshot. On my systems (OS X, Chrome/FF/Safari/Opera, iOS) it looks like Zilla Slab is actually handling Kyrgyz fine - all the glyphs are the same body height: 

![Screen Shot 2020-03-11 at 18 50 14](https://user-images.githubusercontent.com/582683/76472012-ad48e280-63ca-11ea-8e20-1c276aefa3e1.png)

I've compared it to Roboto Slab and the implementations look pretty similar to me, but I could be missing something. Can you tell me what OS/browser you're using?

**Comment by jonorthwash** [comment]
The screenshot above is using Firefox 68.4.2esr in Debian (sid).

The screenshot attached below is using Chromium 80.0.3987.106 on Debian (sid).
![Screenshot from 2020-03-11 23-18-31_cropped](https://user-images.githubusercontent.com/963849/76484015-fca10a00-63ee-11ea-83ea-6c14ade90a28.png)

**Comment by fitojb** [comment]
> it looks like Zilla Slab is actually handling Kyrgyz fine

@phirework, you are actually looking at a fallback font, Georgia 
🤦

**Comment by fitojb** [comment]
Diacritics are broken too in Georgia (fallback font) for Cyrillic. This is on Linux.

![image](https://user-images.githubusercontent.com/554953/80420921-32b42380-88a1-11ea-8e1b-bb3b1bb8ea37.png)

**Comment by jonorthwash** [comment]
> Diacritics are broken too in Georgia (fallback font) for Cyrillic. This is on Linux.
> 
> ![image](https://user-images.githubusercontent.com/554953/80420921-32b42380-88a1-11ea-8e1b-bb3b1bb8ea37.png)

I see multiple <ҫ>s.  Also, Chuvash is properly typed with the unicode characters <ӑ> and <ӗ>—there should be no diacritics.  So you have a text encoding issue compounded by the font issue.

**Comment by fitojb** [comment]
No, it’s UTF-8 — it works properly on copy/paste — but it’s true that there is one Latin cedilla and one Cyrillic (that should be fixed in Pontoon). That is how Skia renders it in Chrome 84. The same happens in Wikipedia. (And I copied Georgia from my Windows 10 partition...) Anyway, Common Voice should ditch the Zilla stack and switch to Inter for Cyrillic locales at least.

**Comment by jonorthwash** [comment]
> No, it’s UTF-8

Of course it's UTF-8.  By "text encoding problem", I had in mind that there should be no unicode diacritics at all—it should just be `U+04D1` and `U+04D7`.  But indeed that's what's used.  Whatever font your system is using is doing a terrible job of rendering those....

For the record, this is what it looks like here:
![Screenshot_2020-04-27 Common Voice by Mozilla](https://user-images.githubusercontent.com/963849/80443449-41b8c700-88dd-11ea-9ca7-bb88bbf8beef.png)

> (that should be fixed in Pontoon)

Lol, good luck with that.  There are [whole language names that are wrong in offensive ways](https://github.com/mozilla/voice-web/issues/2646) in Pontoon.

**Comment by fitojb** [comment]
I found the root of my problem: Chromium is [configured to attempt to use the same font family across a word](https://github.com/chromium/chromium/blob/master/third_party/blink/renderer/platform/fonts/win/font_fallback_win.cc#L110) (going to lenghts like asking HarfBuzz to dynamically render accented letters from existing font glyphs), to avoid the ransom note effect  in your screenshot (which displays Zilla Slab, Georgia and DejaVu Serif as fallback of the fallback), but it makes it worse in this case, because Georgia, as found in Windows 10 Insider (again, this is where I took it), doesn’t contain the combining diacritics required to render Kyrgyz text, so HarfBuzz uses the spacing diacritic glyphs Georgia *does* contain, and that is why they are rendered so far away from their letter.

Firefox instead uses [a cascading approach](https://hg.mozilla.org/mozilla-central/file/tip/gfx/thebes/gfxPlatformFontList.cpp#l1488) to fill in missing letters, which makes the display a bit better but at the cost of three fonts in one paragraph.

</offtopic>

**Comment by phirework** [comment]
Hey folks, can I get some help testing a fix for this on our dev server? Kyrgyz should now be using Noto Serif, which should support the whole expanded alphabet: https://dev.commonvoice.allizom.org/ky 

If this works, it'll be deployed on production in a couple of weeks.

**Comment by jonorthwash** [comment]
> Noto Serif, which should support the whole expanded alphabet:

What's an expanded alphabet?

> If this works, it'll be deployed on production in a couple of weeks.

\o/

**Comment by phirework** [comment]
Sorry, I misspoke. I was thinking of the `latin-extended A` charset, but of course it should also support the Cyrillic and Greek  alphabets.

**Comment by phirework** [comment]
This is finally up on production! Huge thanks to everyone who helped guide this bug over the last -checks watch- three years 🎉

**Comment by jonorthwash** [comment]
I'm still having trouble with the characters ү, ө, and ң in a few places, but not in all browsers on all systems.

![Screenshot from 2021-04-21 19-59-49](https://user-images.githubusercontent.com/963849/115636562-3758bf80-a2dc-11eb-8df8-f896015ba631.png)
![Screenshot from 2021-04-21 19-58-58](https://user-images.githubusercontent.com/963849/115636563-3758bf80-a2dc-11eb-89d2-ab2d88c22a06.png)

This is in Firefox 87.0 on Debian sid.

**Comment by phirework** [comment]
Were you seeing this on the dev server as well or no?

**Comment by jonorthwash** [comment]
Hmm, how can I access these screens on the dev server?

**Comment by phirework** [comment]
Here's the dev server: https://dev.commonvoice.allizom.org/ky 

Also, that last screenshot in particular looks the same to me on both Chrome and Firefox. The [relevant string in Pontoon](https://pontoon.mozilla.org/ky/common-voice/web/locales/en/messages.ftl/?string=193602) is `Common Voice'тун бүгүнкү жаздыруулары`. Is that actually correct? Or should those y's be ү?

**Comment by phirework** [comment]
Here's Noto Serif's glyph map, fwiw - are there any that stand out as missing? https://fonts.google.com/specimen/Noto+Serif#glyphs

**Comment by jonorthwash** [comment]
> Here's the dev server: https://dev.commonvoice.allizom.org/ky

Yes, but the errors are in (1) sentences for validation and (2) the user landing page.  For (1) I went to https://dev.commonvoice.allizom.org/ky/listen and it says there are no sentences to validate.  For (2) I tried logging in and it said I don't have access.

Perhaps restating my question would make it clearer: on the dev server, how can I access the screens that my screenshots are from?

> Also, that last screenshot in particular looks the same to me on both Chrome and Firefox. The [relevant string in Pontoon](https://pontoon.mozilla.org/ky/common-voice/web/locales/en/messages.ftl/?string=193602) is `Common Voice'тун бүгүнкү жаздыруулары`. Is that actually correct? Or should those y's be ү? (I could be entirely misunderstanding where the problem is!)

Yes, that string looks okay.  The issue is that the ө, ү, and ң are different heights than the surrounding characters.


> Here are the glyph maps for [Noto Serif](https://fonts.google.com/specimen/Noto+Serif#glyphs), fwiw (which is used for headings) and [Open Sans](https://fonts.google.com/specimen/Open+Sans#glyphs) (which is used for body) - are there any that stand out as missing?

If I'm understanding this, then both fonts lack all of the following: Өө, Үү, Ңң.  So these are not Kyrgyz-friendly fonts (not to mention not supporting Kazakh, Tatar, Tuvan, Sakha, Halh Mongolian, or quite a few other languages I care about).  Perhaps more inclusive fonts could be considered?

**Comment by phirework** [comment]
Actually, it looks like those are not the complete glyphs for each font, because those characters you listed are all supported: 

https://fonts.google.com/specimen/Noto+Serif?preview.text=%D3%A8%D3%A9,%20%D2%AE%D2%AF,%20%D2%A2%D2%A3&preview.text_type=custom
https://fonts.google.com/specimen/Open+Sans?preview.text=%D3%A8%D3%A9,%20%D2%AE%D2%AF,%20%D2%A2%D2%A3&preview.text_type=custom

I picked those font faces specifically because they had good language supports, so I'm not sure why the problem with different heights is still happening.

**Open Sans**
![Screen Shot 2021-04-24 at 4 42 03 PM](https://user-images.githubusercontent.com/582683/115972478-77a68080-a51c-11eb-9c7b-7a230fd40324.png)

**Noto Serif**
![Screen Shot 2021-04-24 at 4 44 34 PM](https://user-images.githubusercontent.com/582683/115972479-78d7ad80-a51c-11eb-90da-66432bcaec9d.png)

**Comment by fitojb** [comment]
the fonts must’ve been subsetted, then, because they do support those characters

**Comment by phirework** [comment]
> The issue is that the ө, ү, and ң are different heights than the surrounding characters.

I can't tell what you mean with ө, it looks even to me in your screenshot. For ү and ң, do you mean that they shouldn't be descending below the baseline?  

I've also reviewed the imported files for both Open Sans and Noto Serif and they both cover the extended Cyrillic unicode ranges: 

https://fonts.googleapis.com/css?family=Open+Sans:300,400,600,700
https://fonts.googleapis.com/css?family=Noto+Serif:400,700

**Comment by ftyers** [comment]
@phirework it's really slight:

![imatge](https://user-images.githubusercontent.com/449545/117045741-c1a31980-ad07-11eb-9a99-1f0379a34940.png)

But they are slightly lower. Perhaps @jonorthwash could make some font suggestions?

**Comment by jonorthwash** [comment]
> But they are slightly lower. Perhaps @jonorthwash could make some font suggestions?

All the DejaVu fonts seem fine?
